eliot/tests/test_parse.py

```diff
@@ -0,0 +1,167 @@
+"""
+Tests for L{eliot._parse}.
+"""
+
+from __future__ import unicode_literals
+
+from unittest import TestCase
+
+from hypothesis import strategies as st, given, assume
+
+from pyrsistent import PClass, field, pvector_field
+
+from .. import start_action, Message
+from ..testing import MemoryLogger
+from .._parse import Task
+from .._message import WrittenMessage, MESSAGE_TYPE_FIELD, TASK_LEVEL_FIELD
+from .._action import FAILED_STATUS, ACTION_STATUS_FIELD, WrittenAction
+from .strategies import labels
+
+
+class ActionStructure(PClass):
+    """
+    A tree structure used to generate/compare to Eliot trees.
+
+    Individual messages are encoded as a unicode string; actions are
+    encoded as a L{ActionStructure} instance.
+    """
+    type = field(type=(unicode, None.__class__))
+    children = pvector_field(object)  # XXX ("StubAction", unicode))
+    failed = field(type=bool)
+
+    @classmethod
+    def from_written(cls, written):
+        """
+        Create an L{ActionStructure} or L{unicode} from a L{WrittenAction} or
+        L{WrittenMessage}.
+        """
+        if isinstance(written, WrittenMessage):
+            return written.as_dict()[MESSAGE_TYPE_FIELD]
+        else:  # WrittenAction
+            if not written.end_message:
+                raise AssertionError("Missing end message.")
+            return cls(
+                type=written.action_type,
+                failed=(written.end_message.contents[ACTION_STATUS_FIELD] ==
+                        FAILED_STATUS),
+                children=[cls.from_written(o) for o in written.children])
+
+    @classmethod
+    def to_eliot(cls, structure_or_message, logger):
+        """
+        Given a L{ActionStructure} or L{unicode}, generate appropriate
+        structured Eliot log mesages to given L{MemoryLogger}.
+        """
+        if isinstance(structure_or_message, cls):
+            action = structure_or_message
+            try:
+                with start_action(logger, action_type=action.type):
+                    for child in action.children:
+                        cls.to_eliot(child, logger)
+                    if structure_or_message.failed:
+                        raise RuntimeError("Make the eliot action fail.")
+            except RuntimeError:
+                pass
+        else:
+            Message.new(message_type=structure_or_message).write(logger)
+        return logger.messages
```

Comment:
I think this class is OK for this branch, but that the explicit type checks within it point to a possible area for improvement in our objects. 

That is, there should be a common interface shared between `WrittenAction` and `WrittenMessage` (and possibly also their non-written equivalents), or at least something that adapts them to a common interface.